### PR TITLE
[VLM] end2end geo3k multi-turn RL of VLM Recipe

### PR DESCRIPTION
**Pull request by gxlvera**
## Goal
VLM Multi-turn (related to https://github.com/THUDM/slime/issues/1075)

### Rollout
- [x] Created a custom multi-turn rollout function in `examples/vlm_multi_turn/rollout.py`
  - [x] Pluggable interactive env (env path specified via rollout argument --custom-config-path)
  - [x] Early-stop logic
    - [x] `max_turns` (specified via rollout argument --custom-config-path)
    - [x] max_new_token cap
  - [x] `loss_mask` / `rollout_log_probs`
    - `loss_mask = 1` on assistant tokens
    - `loss_mask = 0` on user/observation tokens
    - `rollout_log_probs` padded to match
    - initial `sample.prompt` stays unmasked

### Interactive environment
- [x] Custom env split from rollout: `examples/vlm_multi_turn/env_geo3k.py`
  - `build_env`/ `reset` / `step` / `format_observation` functions for per-turn feedback

### Data & dataset
- [x] Support geo3k multimodal dataset: https://huggingface.co/datasets/chenhegu/geo3k_imgurl
- [ ] OpenCUA not supported yet (may require a more complex interactive env, e.g. https://github.com/xlang-ai/OSWorld)

## Experiment Result
Trained Qwen3-VL-2B-Instruct on the geo3k dataset with multi-turn reasoning, using GRPO. 
final_result

### CODE REVIEW

File: slime/utils/arguments.py [hunk 1]
```diff
@@ -199,6 +199,13 @@ def add_rollout_arguments(parser):
                     "and `truncated`."
                 ),
             )
+            parser.add_argument(
+                "--max-turns",
+                dest="max_turns",
+                type=int,
+                default=None,
+                help="Maximum turns for multi-turn custom rollout (e.g., Sokoban). Defaults to rollout implementation config.",
+            )
```
Comment (zhuzilin) [inline slime/utils/arguments.py]: is it possible to pass these 2 configs through `--custom-config-path`?

Comment (gxlvera) [reply slime/utils/arguments.py]: Yes, it sounds neater. I have pushed the change, thanks!

**Comment by zhaochenyang20** [comment]
Nice done Xiaole!

**Comment by ygxio** [comment]
@gxlvera are you working on OpenCUA part? can i help with it?

**Comment by zhaochenyang20** [comment]
Great job so far!

**Comment by gxlvera** [comment]
> @gxlvera are you working on OpenCUA part? can i help with it?

Hi, you could try to support OpenCUA's AgentNet [dataset](https://huggingface.co/datasets/xlangai/AgentNet). Note that if you want to implement the online interaction, maybe you need an os sandbox for simulation. It's OK if you stick with offline mode (without interaction) although I personally don't think that would work well.

**Comment by ygxio** [comment]
Sure @gxlvera can try that

**Comment by ygxio** [comment]
@gxlvera can you help with openCUA a bit?  where can i connect u?

**Comment by gxlvera** [comment]
> @gxlvera can you help with openCUA a bit? where can i connect u?

Hi, you could DM me at gxlvera@gmail.com~

File: examples/geo3k_vlm_multi_turn/env_geo3k.py [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+        }
+
+        return obs, is_final_turn, info
+
+
+def _extract_ground_truth(sample: Sample | None) -> str | None:
+    """Resolve the ground-truth answer from label or metadata."""
+    if sample is None:
+        return None
+    if getattr(sample, "label", None) is not None:
```
Comment (zhaochenyang20) [inline examples/geo3k_vlm_multi_turn/env_geo3k.py]: Do not use `getattr` as much as possible. Change it to:

`if sample.label is not None:`

**Review by zhaochenyang20 — CHANGES_REQUESTED** [review]

File: examples/geo3k_vlm_multi_turn/rollout.py [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+            continue
+        if isinstance(merged[key], torch.Tensor) and isinstance(val, torch.Tensor):
+            merged[key] = torch.cat([merged[key], val], dim=0)
+        elif isinstance(merged[key], list) and isinstance(val, list):
+            merged[key] = merged[key] + val
+        elif isinstance(merged[key], dict) and isinstance(val, dict):
+            merged[key] = {**merged[key], **val}
+        else:
+            merged[key] = val
+    return merged
```
Comment (zhaochenyang20) [inline examples/geo3k_vlm_multi_turn/rollout.py]: Do not perform torch.cat inside the rollout loop. As the conversation grows, this results in $O(N^2)$ memory allocation and copying overhead. This causes significant latency spikes and GPU memory fragmentation.

Please store the tensors in a list during the loop. Perform a single torch.cat or stack only once when constructing the final Sample object.

File: examples/geo3k_vlm_multi_turn/rollout.py [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+            obs_prompt_ids, obs_image_data, obs_multimodal_inputs, obs_multimodal_train_inputs = (
+                _encode_observation_for_generation(
+                    tokenizer,
+                    processor,
+                    next_user_message,
+                    sample.metadata,
+                    getattr(args, "apply_chat_template", False),
+                    args.apply_chat_template_kwargs,
+                )
+            )
```
Comment (zhaochenyang20) [inline examples/geo3k_vlm_multi_turn/rollout.py]: If possible, cache the processed vision features of the static images and only process the new text/tools delta?

File: examples/geo3k_vlm_multi_turn/rollout.py [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+
+            # Drop a leading BOS if present to avoid injecting it mid-stream.
+            bos_id = getattr(tokenizer, "bos_token_id", None)
```
Comment (zhaochenyang20) [inline examples/geo3k_vlm_multi_turn/rollout.py]: The code relies heavily on getattr(args, ...) and getattr(sample, ...). This defeats static type checking and makes the code fragile (hard to refactor).

File: examples/vlm_multi_turn/rollout.py [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+# When executed as a module: python -m examples.vlm_multi_turn.rollout
+from slime.rollout.sglang_rollout import GenerateState
+from slime.utils.http_utils import post
+from slime.utils.processing_utils import encode_image_for_rollout_engine
+from slime.utils.types import Sample
+
+
+DEFAULT_ENV_MODULE = "examples.vlm_multi_turn.env_geo3k"
+DEFAULT_ROLLOUT_CONFIG = {
+    "max_turns": 5,
```
Comment (zhaochenyang20) [inline examples/vlm_multi_turn/rollout.py]: DEFAULT_ROLLOUT_CONFIG has max_turns=5 in rollout.py but max_turns=3 in env_geo3k.py.

please unify them.

File: examples/geo3k_vlm_multi_turn/env_geo3k.py [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+        if boxed:
+            return str(boxed).strip()
+        for line in reversed(text.splitlines()):
+            cleaned = line.strip()
+            if cleaned:
+                return cleaned[:512]
+        trimmed = text.strip()
+        return trimmed[:512] if trimmed else None
+
+    def _extract_balanced_json(self, text: str, start: int) -> str | None:
```
Comment (zhaochenyang20) [inline examples/geo3k_vlm_multi_turn/env_geo3k.py:125]: Instead of parsing raw strings manually, use a robust library like `dirtyjson` or enforce a strict regex extraction for the <tool_call> block contents. The model should be trained to output valid JSON.

File: examples/geo3k_vlm_multi_turn/rollout.py (hunk 3 above)
Comment (zhaochenyang20) [reply examples/geo3k_vlm_multi_turn/rollout.py]: maybe do it like:

```python
multimodal_inputs_buffer = []

if sample.multimodal_train_inputs:
    multimodal_inputs_buffer.append(sample.multimodal_train_inputs)

for turn_idx in range(max_turns):
    
    if obs_multimodal_train_inputs:
        # only Append, do not Merge
        multimodal_inputs_buffer.append(obs_multimodal_train_inputs)
```

File: examples/geo3k_vlm_multi_turn/rollout.py [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+        elif isinstance(merged[key], list) and isinstance(val, list):
+            merged[key] = merged[key] + val
+        elif isinstance(merged[key], dict) and isinstance(val, dict):
+            merged[key] = {**merged[key], **val}
+        else:
+            merged[key] = val
+    return merged
+
+
+async def generate(args: Any, sample: Sample, sampling_params) -> Sample:
```
Comment (zhaochenyang20) [inline examples/geo3k_vlm_multi_turn/rollout.py]: This function is too huge to review and maintain. Please split it into several parts, like:

1. _prepare_initial_inputs(sample, processor, tokenizer)

 Handles initial encoding.

2. _run_inference_step(url, tokens, sampling_params, image_data, tokenizer)

Handles model inference IO. 

3. _process_env_step(env, env_module, response_text, tokenizer, processor, args, sample_metadata)

Handles environment interaction & observation encoding.

4. _update_multimodal_state(sample, current_image_data, obs_imgs, obs_mm, obs_mm_train)

Handles multimodal state synchronization.

and then, the function is like:

```python
async def generate(args: Any, sample: Sample, params: dict) -> Sample:
    """Minimalist multi-turn rollout orchestration."""
    # 1. Initialization
    env, config, state, url = _initialize_resources(args, sample)
    img_data, budget = _prepare_start_state(sample, state, args, params)

    try:
        env.reset()
        for _ in range(config["max_turns"]):
            
            # --- A. Model Inference (Assistant Turn) ---
            txt, ids, probs, finish = await _run_inference_step(
                url, sample.tokens, params, img_data, state.tokenizer
            )
            
            # Update State: Mask=1 for model outputs (Compute Loss)
            _append_to_sample(sample, ids, probs, loss_mask_val=1)
            
            if _should_stop(sample, finish, budget, len(ids)): break

            # --- B. Environment Interaction (User/System Turn) ---
            obs_ids, obs_imgs, obs_mm, obs_train, done = _process_env_step(
                env, config, txt, state.tokenizer, state.processor, args, sample.metadata
            )
            
            if done: 
                sample.status = Sample.Status.COMPLETED; break

            # Update State: Mask=0 for environment outputs (Ignore Loss)
            _append_to_sample(sample, obs_ids, [0.0]*len(obs_ids), loss_mask_val=0)
            
            # --- C. Sync Multimodal Data ---
            img_data = _update_multimodal_state(sample, img_data, obs_imgs, obs_mm, obs_train)

            if _should_stop(sample, None, budget, len(obs_ids)): break

        return _finalize_sample(sample, state.tokenizer)
        
    finally:
        env.close()
```

File: examples/vlm_multi_turn/rollout.py (hunk 6 above)
Comment (gxlvera) [reply examples/vlm_multi_turn/rollout.py]: Thanks for pointing out. I have unified them. But actually, I personally don't think it's necessary to do so. The logic of setting max_turn is 
1. check user's argument passed as `--custom-config-path`, if None
2. check env's max_turn, if None
3. fall back to the default max_turn set in rollout.py

So it's up to user's input and the specific env's setting.

File: examples/geo3k_vlm_multi_turn/rollout.py (hunk 4 above)
Comment (gxlvera) [reply examples/geo3k_vlm_multi_turn/rollout.py]: What this function does is exactly only process the new content. It's just that the delta content might also include new multimodal data (e.g. for dataset like OpenCUA's GUI-agent trajectories, the env's response is usually a screenshot of the new computer UI). I have cached the multimodal data in previous turns, but we still need to process the new multimodal data (if any) in later turns.

File: examples/vlm_multi_turn/rollout.py (hunk 6 above)
Comment (gxlvera) [reply examples/vlm_multi_turn/rollout.py]: I further discussed with Nan Jiang and decided to delete the default rollout/env config. It's better to force user to input configuration parameter like max_turns for clearer logging and debugging

File: examples/geo3k_vlm_multi_turn/env_geo3k.py (hunk 2 above)
Comment (gxlvera) [reply examples/geo3k_vlm_multi_turn/env_geo3k.py]: Thanks, i have fixed it!

File: examples/geo3k_vlm_multi_turn/rollout.py (hunk 5 above)
Comment (gxlvera) [reply examples/geo3k_vlm_multi_turn/rollout.py]: I have minimized the usage of it, only using it when necessary. In this PR, I now only use getattr() in rollout.py 

```python
def _format_observation(env_module, observation: dict) -> dict:
"""Convert an environment observation into a chat message."""
    formatter = getattr(env_module, "format_observation", None)
    if callable(formatter):
        return formatter(observation)
```

Since it's optional for the environment to provide the formatter, I think it's better to use getattr() here.

File: examples/geo3k_vlm_multi_turn/rollout.py (hunk 3 above)
Comment (gxlvera) [reply examples/geo3k_vlm_multi_turn/rollout.py]: Thanks for the valuable advice! I have implemented the optimized way as below
```python
def _merge_multimodal_train_inputs(chunks: list[dict | None]) -> dict | None:
    """
    Merge per-turn multimodal_train_inputs with a single concat per key.
    """
    if not chunks:
        return None

    values_by_key = {}
    for chunk in chunks:
        if not chunk:
            continue
        for key, val in chunk.items():
            if val is None:
                continue
            values_by_key.setdefault(key, []).append(val)

    merged = {}
    for key, values in values_by_key.items():
        if all(isinstance(v, torch.Tensor) for v in values):
            merged[key] = torch.cat(values, dim=0)

    return merged

for turn_idx in range(max_turns):
    if obs_multimodal_train_inputs:
        multimodal_train_inputs_buffer.append(obs_multimodal_train_inputs)
```

File: examples/geo3k_vlm_multi_turn/env_geo3k.py (hunk 7 above)
Comment (gxlvera) [reply examples/geo3k_vlm_multi_turn/env_geo3k.py:125]: Make sense, i have implemented the robust json extraction as below. (I used orjson instead of dirtyjson is because dirtyjson is not installed by slime's environment) :
```python
def _parse_tool_payload(self, raw_json: str) -> dict[str, Any] | None:
        """Parse tool payload strictly as JSON. Malformed payloads are rejected."""
        loader = orjson.loads
        try:
            return loader(raw_json)
        except Exception as exc:
            logger.warning("Failed to decode tool call payload: %s", exc)
            return None

def _extract_tool_call(self, text: str) -> dict[str, Any] | None:
    #some unmodified codes are skipped
    payload = self._parse_tool_payload(raw_json)
        if payload is None:
            return None
    name = payload.get("name") or payload.get("function", {}).get("name")
    arguments = payload.get("arguments") or payload.get("function", {}).get("arguments") or {}
    if isinstance(arguments, str):
        try:
            arguments = json.loads(arguments)
        except json.JSONDecodeError:
            logger.warning("Tool call arguments are not valid JSON; rejecting tool call.")
            return None
```

File: examples/geo3k_vlm_multi_turn/rollout.py (hunk 8 above)
Comment (gxlvera) [reply examples/geo3k_vlm_multi_turn/rollout.py]: Indeed...😭 I have refactored the [generate function](https://github.com/THUDM/slime/pull/1141/files#diff-120dbcbc06d89becc26b8d53312f6ea7099f5fb015450a6631edaa14c9c94e14)

File: examples/geo3k_vlm_multi_turn/rollout.py (hunk 5 above)
Comment (zhaochenyang20) [reply examples/geo3k_vlm_multi_turn/rollout.py]: getattr can and should be replaced.

Comment (zhaochenyang20) [reply examples/geo3k_vlm_multi_turn/rollout.py]: I eliminated the brittle reliance on getattr and runtime reflection by establishing an explicit OOP contract:

1. Define a BaseInteractionEnv class containing the default format_observation logic and ensure Geo3kEnv inherits from it, which guarantees that the method is always available on the instance and allows me to call env.format_observation() directly.

File: examples/geo3k_vlm_multi_turn/rollout.py (hunk 8 above)
Comment (zhaochenyang20) [reply examples/geo3k_vlm_multi_turn/rollout.py]: Great job

**Comment by zhaochenyang20** [comment]
We shall also have a Megatron version,. But FSDP works cool!

**Comment by zhaochenyang20** [comment]
I evaluated the performance, which works well to me.

image
